server/handlers/handlers.go

```diff
@@ -0,0 +1,108 @@
+package handlers
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/lesterfernandez/roommate-finder/server/data"
+	"github.com/lesterfernandez/roommate-finder/server/token"
+	"golang.org/x/crypto/bcrypt"
+	"net/http"
+)
+
+// ==============================================================================
+//									REGISTER USER
+// ==============================================================================
+
+func RegisterUser(w http.ResponseWriter, res *http.Request) {
+	newUser := data.RegisterBody{}
+	decodeErr := json.NewDecoder(res.Body).Decode(&newUser)
+
+	//if username or password are blank or invalid, throw 400 error
+	if decodeErr != nil || newUser.Username == "" || newUser.Password == "" {
+		respondWithError(w, "Invalid Username or Password", http.StatusBadRequest)
+		fmt.Println(decodeErr)
+		return
+	}
+
+	//if username already exists, throw conflict error
+	if data.UserExists(newUser.Username) {
+		respondWithError(w, "User already exists", http.StatusConflict)
+		return
+	}
+
+	//encrypt password
+	passDigest, hashErr := bcrypt.GenerateFromPassword([]byte(newUser.Password), 10)
+	if hashErr != nil {
+		respondWithError(w, "Something went wrong", http.StatusInternalServerError)
+		return
+	}
+	//temporarily print password, so that its not an unused variable
+	fmt.Printf("Encrypted Password: %v\n", passDigest)
+
+	data.Users = append(data.Users, &newUser)
+
+	jwtToken, signingErr := token.CreateJWT(newUser.Username)
+	if signingErr != nil {
+		respondWithError(w, "Internal Server Error", http.StatusInternalServerError)
```

Comment:
Here we are writing an error to the `ResponseWriter`, but not returning the function. So, if there's a signing error the user will receive an error json string, followed by another json string containing the token. The response wouldn't be valid json and it would be the wrong data.